README.md
![](https://img.shields.io/badge/%20Bioconductor%20BiocCheck-true-green.svg) [![Build Status](https://travis-ci.com/RafalUrniaz/kmeRs.svg?branch=master)](https://travis-ci.com/RafalUrniaz/kmeRs) [![GPLv3 license](https://img.shields.io/badge/License-GPLv3-blue.svg)](http://perso.crans.org/besson/LICENSE.html) [![CRAN_Status_Badge](http://www.r-pkg.org/badges/version/kmeRs)](https://cran.r-project.org/package=kmeRs) [![Codacy Badge](https://api.codacy.com/project/badge/Grade/373b82ef47824c96a0e138afd8e59a02)](https://www.codacy.com/app/RafalUrniaz/kmeRs?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=RafalUrniaz/kmeRs&amp;utm_campaign=Badge_Grade)

# kmeRs: K-Mers Similarity Score Matrix

The package contains tools to calculate similarity score matrix for DNA k-mers. The pairwise similarity score is calculated using PAM or BLOSUM substitution matrix. The results are evaluated by similarity score calculated by Needleman-Wunsch (global) (Needleman and Wunsch 1970) or Smith-Waterman (local) alignment. (Smith and Waterman 1981). Higher similarity score indicates more similar sequences for BLOSUM and less similar sequences for PAM matrix; 30, 40, 70, 120, 250 and 62, 45, 50, 62, 80, 100 matrix versions are available for PAM and BLOSUM, respectively.

## Installation

```r
# Install from CRAN
install.packages("kmeRs")

# Or the development version from GitHub:
# install.packages("devtools")
devtools::install_github("rafalurniaz/kmeRs/kmeRs/")
```
## Documentation and tutorials 
The documentation like manual or tutorials can be found in [documentation](/documentation/) folder to this repository or  directly from installed package by commands:
```r
# To show the introduction vignette
vignette("Introduction_to_kmeRs_package", package="kmeRs")

# To show the manual and functions examples
help(package = "kmeRs")
```
## Quick demo 

```r
kmeRs::kmeRs_test_package()
```

Depends: R (>= 3.4.0) <br/>
Imports: tcR, rDNAse, utils, stats, Biostrings <br/>
Encoding: UTF-8<br/> 

References
----------
###### Needleman, Saul B., and Christian D. Wunsch. 1970. “A General Method Applicable to the Search for Similarities in the Amino Acid Sequence of Two Proteins.” *Journal of Molecular Biology* 48 (3): 443–53. doi:[https://doi.org/10.1016/0022-2836(70)90057-4](https://doi.org/10.1016/0022-2836(70)90057-4).  
###### Smith, T.F., and M.S. Waterman. 1981. “Identification of Common Molecular Subsequences.” *Journal of Molecular Biology* 147 (1): 195–97. doi:[https://doi.org/10.1016/0022-2836(81)90087-5](https://doi.org/10.1016/0022-2836(81)90087-5).



NEWS.md
# kmeRs 1.1.0 - CRAN available  

- Main package functions (added, tested, submitted, and published): 

* `kmeRs_similarity_matrix()`
* `kmeRs_score_and_sort()`
* `kmeRs_show_alignment()`
* `kmeRs_statistics()`
* `kmeRs_transcript_to_complementary()`
* `kmeRs_test_package()`

kmeRs/R/kmeRs_similarity_matrix.R
#' @title Pairwise Similarity Matrix
#'
#' @description
#' The \code{kmeRs_similarity_matrix} function generates a pairwise similarity score
#' matrix for for k length given k-mers vs. all possible k-mers combination.
#' The pairwise similarity score is calculated using PAM or BLOSUM substitution matrix;
#' 30, 40, 70, 120, 250 and 62, 45, 50, 62, 80, 100 matrix versions are available for
#' PAM or BLOSUM, respectively. The results are evaluated by global similarity score;
#' higher similarity score indicates more similar sequences for BLOSUM and opposite for
#' PAM matrix.
#'
#' @aliases kmeRs_similarity_matrix
#'
#' @param kmers_given vector with given k-mers
#' @param compare_to this parameter can have 3 different states, when:
#' '' - the \code{kmers_given} will be compared to each other, default value;
#' ALL -  the \code{kmers_given} will be compared to all possible combinations given by k parameter e.g. N= 4^6 = 4096 combinations for 6-mers;
#' 3rd option is to provide a list of k-mers which should be compared with the set given by the \code{kmers_given} parameter
#' @param k length of k-mers to calculate similarity matrix, higher values may slow down the computer, default value is k=3
#' @param alignment_type type of alignment, default is 'global', could be 'local' or 'global', where 'global' represents
#' Needleman-Wunsch global alignment; 'local' represents Smith-Waterman local alignment.
#' @param submat substitution matrix, default is 'BLOSUM62', but could be one of 'BLOSUM45', 'BLOSUM50', 'BLOSUM62', 'BLOSUM80',
#' 'BLOSUM100', 'PAM30', 'PAM40', 'PAM70', 'PAM120', 'PAM250'
#' @param save_to_file directory and file name; if value is declared the matrix will be saved to the given file name
#'
#' @return similarity matrix is returned as a data.frame
#'
#' @examples
#' # Display BLOSUM matrix used for calculation
#'
#' kmeRs_similarity_matrix(kmers_given = c("A", "T", "C", "G"), submat = "BLOSUM62")
#'
#' @importFrom rDNAse twoSeqSim
#' @importFrom utils write.csv2
#' @importFrom tcR generate.kmers
#' @importFrom Biostrings score
#'
#' @export


 kmeRs_similarity_matrix <- function(kmers_given, compare_to = '', alignment_type = "global", k = 3, submat = "BLOSUM62", save_to_file = '' ){

  # Avoid case-sensitive errors

    kmers_given <- toupper(kmers_given)

  # Compare to the same set by default

    if (compare_to[1] == ''){

      compare_to <- kmers_given

    }

  # If compare_to_all == TRUE - Generate all possible k-mers from DNA alphabet

    if (compare_to[1] == 'ALL'){

      compare_to <- tcR::generate.kmers(.k = k, .seq = '', .alphabet = c('A', 'C', 'G', 'T'))

    }

  ## -- Prepare matrix --

    kmers_dist_matrix <- matrix(NA, ncol = length(kmers_given), nrow = length(compare_to))

  # Set cols and rows names

    kmers_dist_matrix <- data.frame(kmers_dist_matrix, row.names = compare_to)
    colnames(kmers_dist_matrix) <- kmers_given


  ## -- Calculate the distance matrix -- TO-DO: Optimise performance - parallel calculations, packages like doParallel and foreach

    for (col in 1:length(kmers_dist_matrix[1, ])){

      for (row in 1:length(kmers_dist_matrix[, 1])){

        seqA <- colnames(kmers_dist_matrix)[col]
        seqB <- rownames(kmers_dist_matrix)[row]

        kmers_dist_matrix[row, col] <- Biostrings::score(rDNAse::twoSeqSim(seqA, seqB, type = "global", submat = submat))  # TO-DO: more precise method may be applied

      }
    }

  # Save to the file if requested

  if (save_to_file > ''){

    utils::write.csv2(kmers_dist_matrix, save_to_file)

  }

  # Return matrix

    return(kmers_dist_matrix)

}
